ee/backend/app/src/main/kotlin/io/tolgee/ee/service/prompt/PromptServiceEeImpl.kt

```diff
@@ -0,0 +1,461 @@
+        jacksonObjectMapper().readValue<JsonNode>(result.response)
+      } catch (e: JsonProcessingException) {
+        null
+      }
+    return result
+  }
+
+  fun runPrompt(
+    organizationId: Long,
+    params: LlmParams,
+    provider: String,
+    attempts: List<Int>? = null,
+  ): PromptResult {
+    publishBeforeEvent(organizationId)
+    val result = runPromptWithoutChargingCredits(organizationId, params, provider, attempts)
+    publishAfterEvent(organizationId, result.price)
+    return result
+  }
+
+  fun <T> rateLimitToStatus(func: () -> T): T {
+    return try {
+      func()
+    } catch (e: BadRequestException) {
+      if (e.tolgeeMessage == Message.LLM_RATE_LIMITED) {
+        throw TooManyRequestsException(Message.TOO_MANY_REQUESTS, cause = e)
+      } else {
+        throw e
+      }
+    }
+  }
+
+  fun getTranslationFromPromptResult(result: PromptResult): MtValueProvider.MtResult {
+    val json = result.parsedJson ?: throw BadRequestException(Message.LLM_PROVIDER_NOT_RETURNED_JSON)
+    val translation = json.get("output")?.asText() ?: throw BadRequestException(Message.LLM_PROVIDER_NOT_RETURNED_JSON)
+
+    return MtValueProvider.MtResult(
+      translation,
+      contextDescription = json.get("contextDescription")?.asText(),
+      price = result.price,
+      usage = result.usage,
+    )
+  }
+
+  @Transactional
+  override fun translate(
+    projectId: Long,
+    data: PromptRunDto,
+    priority: LlmProviderPriority?,
+  ): MtValueProvider.MtResult {
+    val project = projectService.get(projectId)
+    val prompt =
+      getPrompt(
+        projectId,
+        data.template ?: promptDefaultService.getDefaultPrompt().template!!,
+        data.keyId,
+        data.targetLanguageId,
+        data.provider,
+        data.options,
+      )
+    val params = getLlmParamsFromPrompt(prompt, data.keyId, priority ?: LlmProviderPriority.HIGH)
+    val result = runPrompt(project.organizationOwner.id, params, data.provider)
+    return getTranslationFromPromptResult(result)
+  }
+
+  override fun translateAndUpdateTranslation(
+    projectId: Long,
+    data: PromptRunDto,
+    priority: LlmProviderPriority?,
+  ) {
+    val result = translate(projectId, data, priority)
+    val translation = translationService.getOrCreate(data.keyId, data.targetLanguageId)
+    translation.text = result.translated
+    translationService.save(translation)
+  }
+
+  fun getDefaultPrompt(): PromptDto {
+    return promptDefaultService.getDefaultPrompt()
+  }
+
+  private fun publishBeforeEvent(organizationId: Long) {
+    applicationContext.publishEvent(
+      OnBeforeMachineTranslationEvent(this, organizationId),
+    )
+  }
+
+  private fun publishAfterEvent(
+    organizationId: Long,
+    actualPriceInCents: Int,
+  ) {
+    applicationContext.publishEvent(
+      OnAfterMachineTranslationEvent(this, organizationId, actualPriceInCents),
+    )
+  }
+
+  companion object {
+    class LazyMap : AbstractMap<String, Any?>() {
```

Comment:
This file is already large enough to contain inner classes in it. Move it to separate files. This can be useful to use in general.